// next-sanity/dist/studio.cjs
"use strict";
Object.defineProperty(exports, "__esModule", { value: !0 });
var NextStudioNoScript = require("./_chunks-cjs/NextStudioNoScript.cjs"), jsxRuntime = require("react/jsx-runtime"), clientComponent = require("next-sanity/studio/client-component"), reactDom = require("react-dom");
const viewport = {
  width: "device-width",
  initialScale: 1,
  // Studio implements display cutouts CSS (The iPhone Notch ™ ) and needs `viewport-fit=covered` for it to work correctly
  viewportFit: "cover"
}, metadata = {
  referrer: "same-origin",
  robots: "noindex"
}, bridgeScript = "https://core.sanity-cdn.com/bridge.js";
function NextStudioWithBridge(props) {
  return reactDom.preloadModule(bridgeScript, { as: "script" }), /* @__PURE__ */ jsxRuntime.jsxs(jsxRuntime.Fragment, { children: [
    /* @__PURE__ */ jsxRuntime.jsx("script", { src: bridgeScript, async: !0, type: "module", "data-sanity-core": !0 }),
    /* @__PURE__ */ jsxRuntime.jsx(clientComponent.NextStudio, { ...props })
  ] });
}
exports.NextStudioLayout = NextStudioNoScript.NextStudioLayout;
exports.NextStudioNoScript = NextStudioNoScript.NextStudioNoScript;
exports.NextStudio = NextStudioWithBridge;
exports.metadata = metadata;
exports.viewport = viewport;
